docs/proposals/04-etcd-member-custom-resource.md

```diff
@@ -0,0 +1,561 @@
+
+```yaml
+apiVersion: druid.gardener.cloud/v1alpha1
+kind: EtcdMember
+metadata:
+  labels:
+    gardener.cloud/owned-by: <name of parent Etcd resource>
+  name: <name of the etcd-member>
+  namespace: <namespace | will be the same as that of parent Etcd resource>
+  ownerReferences:
+  - apiVersion: druid.gardener.cloud/v1alpha1
+    blockOwnerDeletion: true
+    controller: true
+    kind: Etcd
+    name: <name of the parent Etcd resource>
+    uid: <UID of the parent Etcd resource> 
+status:
+  id: <etcd-member id>
+  clusterID: <etcd cluster id>
+  peerTLSEnabled: <bool>
+  dbSize: <db-size>
+  dbSizeInUse: <db-size-in-use>
+  snapshots:
+    lastFull:
+      timestamp: <time of full snapshot>
+      name: <name of the file that is uploaded>
+      size: <size of the un-compressed snapshot file uploaded>
+      startRevision: <start revision of etcd db captured in the snapshot>
+      endRevision: <end revision of etcd db captured in the snapshot>
+    lastDelta:
+      timestamp: <time of delta snapshot>
+      name: <name of the file that is uploaded>
+      size: <size of the un-compressed snapshot file uploaded>
+      startRevision: <start revision of etcd db captured in the snapshot>
+      endRevision: <end revision of etcd db captured in the snapshot>
+    accumulatedDeltaSize: <total size of delta snapshots since last full snapshot>
+  restorations:
+  - type: <FromSnapshot | FromLeader>
+    status: <Failed | Success | In-Progress>
+    startTime: <start time of restoration>
+    endTime: <end time of restoration>
+  defragmentations:
+  - startTime: <start time of defragmentation>
+    endTime: <end time of defragmentation>
+    reason: 
+    message:
+    initialDBSize: <size of etcd DB prior to defragmentation>
+    finalDBSize: <size of etcd DB post degragmentation>
+  volumeMismatches:
+  - identifiedAt: <time at which wrong volume mount was identified>
+    fixedAt: <time at which correct volume was mounted>
+    volumeID: <volume ID of wrong volume that got mounted>
+    numRestarts: <num of pod restarts that were attempted>
+  transitions:
+  - state: <name of the state that the etcd-member has transitioned to>
+    subState: <name of the sub-state if any>
+    reason: <reason code for the transition>
+    transitionTime: <time of transition to this state>
+    message: <detailed message if any>
+```
+
+### Lifecycle of an EtcdMember
+
+#### Creation
+
+Druid creates an `EtcdMember` resource for every replica in `etcd.Spec.Replicas` during reconciliation of an etcd resource. For a fresh etcd cluster this is done prior to creation of the StatefulSet resource and for an existing cluster which has now been scaled-up, it is done prior to updating the StatefulSet resource.
+
+#### Updation
+
+All fields in `EtcdMember.Status` are *only* updated by the corresponding etcd-member. Druid only consumes the information published via `EtcdMember` resources.
```

Comment:
If we are going ahead with the status updation for a myriad number of reasons, a single Druid controller requires to orchestrate multiple of things. What worries me is suppose the druid controller is orchestrating defragmentation of a member and at the same time status is updated for restoration for another member. Can druid handle both the scenarios?

Reply:
So lets take case by case. 
Take a case that defragmentation is started. Typically as it happens today the current leader will be the last one to be selected for defragmentation to prevent multiple leader elections during the entire process. Let us assume that pod-0 is the current leader and let us assume that pod-1 and pod-2 are heathy and there is cluster readiness (quorum is established).

Druid picks up pod-1, while pod-1 is defragmenting there is a leadership change and the new leader is now pod-2. When pod-1's defragmentation finishes, then druid will re-inspect the cluster state. If there is a quorum and there are still pending members to be defragmented then druid will choose any non-leader that has not yet been defragmented. It will requested that member to initiate the defragmentation. Since druid does defragmentation one member at a time, it will always get the latest state of each member.

Now consider a case that there was a leadership change but the status update got delayed. Max what will happen is that it will cause a re-election. We will need to observe how frequent would this be (leadership change) in real world.

Can you list down other cases and we can go over them and see which cases are problematic.

Reply:
What I am suggesting here is let Druid not handle the defragmentation or restoration. I think it would be less complex and easier to debug in case of fault if individual members handle these. I am still not buying the fact that druid should handle everything for ETCD BR. Let's not increase the complexities and make things difficult to debug and resolve. It would be a real problem, in case of a failure, to pore over the logs of druid and figure out which steps happened when and after what status change.

Reply:
>What I am suggesting here is let Druid not handle the defragmentation or restoration. I think it would be less complex and easier to debug in case of fault if individual members handle these

On the contrary, this is much more complex. A sidecar should only serve its respective main container within the pod and not engage in cluster wide operations like defragmentations. Reasons:
1. Leadership can change at any given point of time.
2. Any member can go down at any given point of time.
Due to above two reasons its important that the defragmentation be stopped. Its actually quite easy to implement, maintain and orchestrate when you do this from druid. It is not only simple but it is completely transparent of the state of defragmentation for the entire cluster.

>Let's not increase the complexities and make things difficult to debug and resolve. It would be a real problem

I would like to understand the reason why you think it is more complex and what is difficult to debug. This statement is quite generic and i would not be able to respond properly unless i understand what problems you see.

Reply:
In the meeting on 23rd Aug we agreed on the following:
During the implementation, care should be taken that only one actor makes changes to the `EtcdMember` resource in backup-restore container. This prevents issues of conflict during updates. It is also understood that there is always a possibility that these updates get delayed or fail. These situations should be properly handled.